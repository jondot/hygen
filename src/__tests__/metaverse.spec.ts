import Logger from '../logger'

jest.mock('enquirer', () => ({
  prompt: null,
}))

const SKIP_ON_WINDOWS = process.platform === 'win32' ? ['shell'] : []

const path = require('path')
const dirCompare = require('dir-compare')

const opts = {
  compareContent: true,
  compareFileSync:
    dirCompare.fileCompareHandlers.defaultFileCompare.compareSync,
  compareFileAsync:
    dirCompare.fileCompareHandlers.defaultFileCompare.compareAsync,
  ignoreLineEnding: true,
  ignoreWhiteSpaces: true
}
const fs = require('fs-extra')
const enquirer = require('enquirer')
const { runner } = require('../index')

const logger = new Logger(console.log)
const failPrompt = () => {
  throw new Error('set up prompt in testing')
}

const dir = m => path.join(__dirname, 'metaverse', m)
const metaverse = (folder, cmds, promptResponse = null) =>
  it(folder, async () => {
    const metaDir = dir(folder)
    console.log('metaverse test in:', metaDir)
    const config = {
      templates: '_templates',
      cwd: metaDir,
      exec: (action, body) => {
        const opts = body && body.length > 0 ? { input: body } : {}
        return require('execa').shell(action, opts)
      },
      logger,
      createPrompter: () => require('enquirer'),
    }
    // await fs.remove(path.join(metaDir, 'given'))
    console.log('before', fs.readdirSync(metaDir))
    for (let cmd of cmds) {
      console.log('testing', cmd)
      if (
        process.platform === 'win32' &&
        SKIP_ON_WINDOWS.find(c => cmd[0] === c)
      ) {
        console.log(`skipping ${cmd} (windows!)`)
        await fs.remove(path.join(metaDir, 'expected', cmd[0]))
        continue
      }

      enquirer.prompt = failPrompt
      if (promptResponse) {
        const last = cmd[cmd.length - 1]
        if (typeof last === 'object') {
          cmd = cmd.slice(cmd.length - 1)
          enquirer.prompt = () =>
            Promise.resolve({ ...promptResponse, ...last })
        } else {
          enquirer.prompt = () => Promise.resolve(promptResponse)
        }
      }
      const res = await runner(cmd, config)
      res.actions.forEach(a => {
        a.timing = -1
        a.subject = a.subject.replace(/.*hygen\/src/, '')
      })
      expect(res).toMatchSnapshot(cmd.join(' '))
    }
    const givenDir = path.join(metaDir, 'given')
    const expectedDir = path.join(metaDir, 'expected')
    console.log('after', {
      [givenDir]: fs.readdirSync(givenDir),
      [expectedDir]: fs.readdirSync(expectedDir),
    })
    const res = dirCompare.compareSync(givenDir, expectedDir, opts)
    res.diffSet = res.diffSet.filter(d => d.state !== 'equal')
    if (!res.same) {
      console.log(res)
    }
    expect(res.same).toEqual(true)
  })

describe('metaverse', () => {
  beforeEach(() => {
    enquirer.prompt = failPrompt
  })
  metaverse('hygen-extension', [['hygen-js', 'new']], { overwrite: true })
  metaverse(
    'hygen-templates',
    [
      ['init', 'self'],
      ['overwrite-yes', 'base'],
      ['overwrite-yes', 'over'],
      ['overwrite-no', 'base'],
      ['overwrite-no', 'over', { overwrite: false }],
      ['mailer', 'new'],
      ['worker', 'new', '--name', 'foo'],
      ['shell', 'new', '--name', 'foo'],
      ['inflection', 'new', '--name', 'person'],
      ['conditional-rendering', 'new', '--notGiven'],
      ['add-unless-exists', 'new', '--message', 'foo'],
      [
        'cli-prefill-prompt-vars',
        'new',
        '--message-from-cli',
        'hello-from-cli',
      ],
      ['cli-prefill-prompt-vars', 'name-is-special', 'foobar'],
      [
        'cli-prefill-prompt-vars',
        'falsy-values-are-ok',
        'foobar',
        '--include_something',
        'false',
      ],
      ['recursive-prompt', 'new'],
      ['positional-name', 'new', 'acmecorp'],
      ['existing-params', 'new', '--email', 'premade-email@foobar.com'],
      [
        'existing-params',
        'new-params-alias',
        '--email',
        'premade-email@foobar.com',
      ],
      [
        'index-js-existing-params',
        'new',
        '--email',
        'premade-email@foobar.com',
      ],
      [
        'index-js-existing-params',
        'new-params-alias',
        '--email',
        'premade-email@foobar.com',
      ],
    ], // this is all of the responses enquirer gives out from _all_ tests, ever.
    // it's best to just keep it that way to be simple, and each prompt-dealing test
    // has its own set of uniquely named variables.
    {
      // generic for all tests
      name: 'message',
      message: 'foo',
      overwrite: true,

      // recursive-prompt
      email: 'some-email@foobar.com',
      emailConfirmation: 'confirmed-some-email@foobar.com',
    },
  )
})
