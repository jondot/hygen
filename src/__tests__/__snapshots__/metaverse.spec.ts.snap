// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metaverse hygen-extension: hygen-js new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/hygen-js/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: [object Object] 1`] = `
Object {
  "actions": Array [],
  "success": false,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: add-unless-exists new --message foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "added",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "skipped",
      "subject": "given/add-unless-exists/foo/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: cli-prefill-prompt-vars falsy-values-are-ok foobar --include_something false 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/cli-prefill-prompt-vars/falsy.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: cli-prefill-prompt-vars name-is-special foobar 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/cli-prefill-prompt-vars/name-is-special.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: cli-prefill-prompt-vars new --message-from-cli hello-from-cli 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/cli-prefill-prompt-vars/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: conditional-rendering new --notGiven 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/conditional/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: existing-params new --email premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/existing-params/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: existing-params new-params-alias --email premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/existing-params/new-params-alias.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: index-js-existing-params new --email premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/index-js-existing-params/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: index-js-existing-params new-params-alias --email premade-email@foobar.com 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/index-js-existing-params/new-params-alias.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: inflection new --name person 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/my_app/people.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: init self 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/_templates/sample/new/hello.ejs.t",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: mailer new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/app/mailers/message.js",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "added",
      "subject": "given/app/mailers/foo/html.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "added",
      "subject": "given/app/mailers/foo/subject.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "added",
      "subject": "given/app/mailers/foo/text.ejs",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "inject",
      "subject": "given/app/mailers/foo/html.ejs",
      "timing": -1,
      "type": "inject",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: overwrite-no base 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/overwrite-no/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: overwrite-yes base 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/overwrite-yes/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: overwrite-yes over 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/overwrite-yes/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: positional-name new acmecorp 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/positional-name/acmecorp/always.txt",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: recursive-prompt new 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/recursive-prompt/new.md",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: shell new --name foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/hygen-templates/given/shell/hello.piped",
      "timing": -1,
      "type": "shell",
    },
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/hygen-templates/given/shell/foo.hello",
      "timing": -1,
      "type": "shell",
    },
    Object {
      "status": "added",
      "subject": "given/shell/hmm.txt",
      "timing": -1,
      "type": "add",
    },
    Object {
      "status": "executed",
      "subject": "/__tests__/metaverse/hygen-templates/given/shell/side-effect.hello",
      "timing": -1,
      "type": "shell",
    },
  ],
  "success": true,
  "time": 0,
}
`;

exports[`metaverse hygen-templates: worker new --name foo 1`] = `
Object {
  "actions": Array [
    Object {
      "status": "added",
      "subject": "given/app/workers/foo.js",
      "timing": -1,
      "type": "add",
    },
  ],
  "success": true,
  "time": 0,
}
`;
